### ISSUE TITLE
pam_callback! triggers clippy::not_unsafe_ptr_arg_deref

### ISSUE THREAD

**Issue by carnott-snap** [rcatolino/pam_sm_rust#12]
The following code, from your docs, builds as expected:
```rust
#[macro_use] extern crate pamsm;
extern crate time;

use pamsm::{PamServiceModule, Pam, PamFlag, PamError};

struct PamTime;

impl PamServiceModule for PamTime {
    fn authenticate(pamh: Pam, _: PamFlag, args: Vec<String>) -> PamError {
        let hour = time::now().tm_hour;
        if hour != 4 {
            // Only allow authentication when it's 4 AM
            PamError::SUCCESS
        } else {
            PamError::AUTH_ERR
        }
    }
}

pam_module!(PamTime);
```
But, causes `clippy::not_unsafe_ptr_arg_deref` to blow up:
```
error: this public function dereferences a raw pointer but is not marked `unsafe`
  --> src/lib.rs:20:1
   |
20 | pam_module!(PamTime);
   | ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(clippy::not_unsafe_ptr_arg_deref)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#not_unsafe_ptr_arg_deref
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```
This happens six times, so it seems pretty clear that the issue is contained within [`pam_callback!`](https://github.com/rcatolino/pam_sm_rust/blob/ef541bfd1b92d0276a3d2cbc42a6c9f1b5ad9c19/src/pam.rs#L246), but I did not think that `pub extern "C" fn` also needed to be tagged as `unsafe`, or am I missing something?

**Comment by lucab** [comment]
@carnott-snap I agree with your analysis. We can mark as `unsafe` on the macro side or silence the warning internally. However the question below is interesting:

> I did not think that pub extern "C" fn also needed to be tagged as unsafe, or am I missing something?

This came to my mind a few times, but I was always unsure. It seems you are sharing the same sentiment, so it may be a good time to ask clippy developers if they agree on this and if they can consider a non-erroring corner-case for this. Would you mind opening a ticket there, and keep us in the loop?

**Comment by rcatolino** [comment]
I'm not opposed marking the callbacks as unsafe. After all, if the caller passes an invalid pointer for ```argv```, the callback is very much unsafe.

Of course the fact that's it's only meant to be called from C code makes the ```unsafe``` kinda pointless, but not semantically wrong.

Can you see any downsides marking the callbackss ```unsafe``` ?

**Comment by carnott-snap** [comment]
> Of course the fact that's it's only meant to be called from C code makes the `unsafe` kinda pointless, but not semantically wrong.

To be clear, I am purely interested in making clippy happy, but I agree with the sentiment.

**Comment by rcatolino** [comment]
I've marked the callbacks as unsafe and published the new version as 0.3.1
@lucab I've also invited you as an owner on crates.io because I forgot last time.
